### ISSUE TITLE
Invalid value `(nil)`  at index 1 for table `concat`

### ISSUE THREAD

**Issue by Frityet** [luarocks/luarocks#1460]
### Please read the following before submitting:
- Please do NOT submit bug reports for questions. Ask questions in the [Gitter
  chat-room](https://gitter.im/luarocks/luarocks).
- Please do NOT submit duplicate bug reports. Look through the [issue
  tracker](https://github.com/luarocks/luarocks/issues) before submitting.

### Please fill out the following:
- **Platform:**
  - Windows/Linux/MacOS.
 MacOS Monterey x86-64

- **LuaRocks version:**
  - `luarocks --version`
 3.9.1

- **Configuration file:**
  - Please try to produce with the default configuration.
  - If you cannot reproduce with the default configuration, please try to find
    the minimal configuration to reproduce.
  - Upload the config to a pastebin such as gist.github.com.
 ```yaml
lua_interpreter = "lua5.4"
variables = {
   LUA_BINDIR = "/usr/local/opt/lua/bin",
   LUA_DIR = "/usr/local/opt/lua",
   LUA_INCDIR = "/usr/local/opt/lua/include/lua5.4"
}
```

- **LuaRocks output from when the issue occurred:**
  - Use the `--verbose` flag.

```yaml
Defaulting to Lua 5.4 based on /Users/frityet/Documents/Fuser/.luarocks/default-lua-version.lua ...
fs.current_dir()
fs.current_dir()
fs.is_superuser()
fs.current_user()
fs.exists("test/luarocks-build-fuser-test-dev-1.rockspec")
fs.absolute_name("test/luarocks-build-fuser-test-dev-1.rockspec")
fs.current_dir()
fs.absolute_name("test/luarocks-build-fuser-test-dev-1.rockspec")
fs.current_dir()
fs.check_command_permissions(table: 0x7fa2aea86130)
fs.exists("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4")
fs.is_writable("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4")
fs.is_dir("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4")
fs.absolute_name("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4")
fs.current_dir()
fs.exists("/Users/frityet/Documents/Fuser/lua_modules/share/lua/5.4")
fs.is_writable("/Users/frityet/Documents/Fuser/lua_modules/share/lua/5.4")
fs.is_dir("/Users/frityet/Documents/Fuser/lua_modules/share/lua/5.4")
fs.absolute_name("/Users/frityet/Documents/Fuser/lua_modules/share/lua/5.4")
fs.current_dir()
fs.exists("/Users/frityet/Documents/Fuser/lua_modules/bin")
fs.is_writable("/Users/frityet/Documents/Fuser/lua_modules/bin")
fs.is_dir("/Users/frityet/Documents/Fuser/lua_modules/bin")
fs.absolute_name("/Users/frityet/Documents/Fuser/lua_modules/bin")
fs.current_dir()
fs.exists("/Users/frityet/Documents/Fuser/lua_modules/share/lua/5.4")
fs.is_writable("/Users/frityet/Documents/Fuser/lua_modules/share/lua/5.4")
fs.is_dir("/Users/frityet/Documents/Fuser/lua_modules/share/lua/5.4")
fs.absolute_name("/Users/frityet/Documents/Fuser/lua_modules/share/lua/5.4")
fs.current_dir()

luarocks-build-fuser-test dev-1 depends on lua ~> 5, >= 5.1 (5.4-1 provided by VM)
luarocks-build-fuser-test dev-1 depends on lua-cjson (2.1.0.10-1 installed)
fs.is_dir("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1")
fs.absolute_name("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1")
fs.current_dir()
fs.make_dir("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/lib")
fs.set_permissions("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test", "exec", "all")
fs._unix_moderate_permissions("755")
fs._unix_umask()

io.popen:       umask -S
fs.execute("chmod", "755", "/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test")
fs.Q("755")
fs.Q("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test")
fs.execute_string("chmod '755' '/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test'")

os.execute:     chmod '755' '/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test'
Results: 3
  1 (boolean): true
  2 (string): exit
  3 (number): 0
fs.set_permissions("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1", "exec", "all")
fs._unix_moderate_permissions("755")
fs._unix_umask()
fs.execute("chmod", "755", "/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1")
fs.Q("755")
fs.Q("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1")
fs.execute_string("chmod '755' '/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1'")

os.execute:     chmod '755' '/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1'
Results: 3
  1 (boolean): true
  2 (string): exit
  3 (number): 0
fs.set_permissions("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/lib", "exec", "all")
fs._unix_moderate_permissions("755")
fs._unix_umask()
fs.execute("chmod", "755", "/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/lib")
fs.Q("755")
fs.Q("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/lib")
fs.execute_string("chmod '755' '/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/lib'")

os.execute:     chmod '755' '/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/lib'
Results: 3
  1 (boolean): true
  2 (string): exit
  3 (number): 0
fs.make_dir("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/conf")
fs.set_permissions("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/conf", "exec", "all")
fs._unix_moderate_permissions("755")
fs._unix_umask()
fs.execute("chmod", "755", "/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/conf")
fs.Q("755")
fs.Q("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/conf")
fs.execute_string("chmod '755' '/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/conf'")

os.execute:     chmod '755' '/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/conf'
Results: 3
  1 (boolean): true
  2 (string): exit
  3 (number): 0
fs.make_dir("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/bin")
fs.set_permissions("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/bin", "exec", "all")
fs._unix_moderate_permissions("755")
fs._unix_umask()
fs.execute("chmod", "755", "/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/bin")
fs.Q("755")
fs.Q("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/bin")
fs.execute_string("chmod '755' '/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/bin'")

os.execute:     chmod '755' '/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/bin'
Results: 3
  1 (boolean): true
  2 (string): exit
  3 (number): 0
fs.make_dir("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/lua")
fs.set_permissions("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/lua", "exec", "all")
fs._unix_moderate_permissions("755")
fs._unix_umask()
fs.execute("chmod", "755", "/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/lua")
fs.Q("755")
fs.Q("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/lua")
fs.execute_string("chmod '755' '/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/lua'")

os.execute:     chmod '755' '/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1/lua'
Results: 3
  1 (boolean): true
  2 (string): exit
  3 (number): 0

Error: LuaRocks 3.9.1 bug (please report at https://github.com/luarocks/luarocks/issues).
Arch.: macosx-x86_64
...ellar/luarocks/3.9.1/share/lua/5.4/luarocks/core/dir.lua:33: invalid value (nil) at index 1 in table for 'concat'
stack traceback:
        [C]: in function 'table.concat'
        ...ellar/luarocks/3.9.1/share/lua/5.4/luarocks/core/dir.lua:33: in function 'luarocks.dir.path'
        ...llar/luarocks/3.9.1/share/lua/5.4/luarocks/core/path.lua:18: in function 'luarocks.core.path.rocks_dir'
        lua_modules/share/lua/5.4/luarocks/build/fuser.lua:16: in function 'luarocks.build.fuser.run'
        ...l/Cellar/luarocks/3.9.1/share/lua/5.4/luarocks/build.lua:208: in upvalue 'run_build_driver'
        ...l/Cellar/luarocks/3.9.1/share/lua/5.4/luarocks/build.lua:412: in function 'luarocks.build.build_rockspec'
        ...ellar/luarocks/3.9.1/share/lua/5.4/luarocks/cmd/make.lua:130: in function 'luarocks.cmd.make.command'
        (...tail calls...)
        [C]: in function 'xpcall'
        ...cal/Cellar/luarocks/3.9.1/share/lua/5.4/luarocks/cmd.lua:672: in function 'luarocks.cmd.run_command'
        ...cal/lib/luarocks/rocks-5.1/luarocks/3.9.1-1/bin/luarocks:35: in main chunk
        [C]: in ?
fs.change_dir_to_root()
fs.delete("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test/dev-1")
fs.remove_dir_if_empty("/Users/frityet/Documents/Fuser/lua_modules/lib/luarocks/rocks-5.4/luarocks-build-fuser-test")
```

- **Description:**
  - The steps you took in plain English to reproduce the problem.

I am treying to test my build plugin, [fuser](https://github.com/Frityet/Fuser/tree/master) using the command
```sh
./luarocks --verbose make test/luarocks-build-fuser-test-dev-1.rockspec  
```
where `./luarocks` was modified to also search in my local `lua_modules`.

**Comment by Frityet** [comment]
what is the proper procedure for testing these types of plugins anyways?

**Comment by Frityet** [comment]
@hishamhm why was the bug label removed? Was it fixed or user error?

**Comment by hishamhm** [comment]
@Frityet I removed the label because this is not confirmed as a bug in LuaRocks. Crashes during development as an extension can happen due to misuse of internal APIs (in that case it would be more similar to "user error" than a bug). To your other question about the procedure for testing these extensions, indeed we don't have a good user story for it — there hasn't been much interest in writing extensions historically, so support is pretty minimal at this time, unfortunately. These extensions points are provided in a best-effort basis, but sadly we don't have a documented stable programmatic API or testing tooling to point you to.

If you require `luarocks.*` modules from your extension and pass invalid arguments to functions, you will get crashes, and LuaRocks will say "please report a bug" along with the stacktrace because that's the default behavior of the custom error handler designed to catch LuaRocks internal bugs — I agree that's misleading, but there's not much we can do there.

**Comment by hishamhm** [comment]
(also, the "bug" label was originally intended for triaged/confirmed LuaRocks bugs, but the issue template was auto-adding it for every reported issue — I have since fixed that)

**Comment by Frityet** [comment]
> @Frityet I removed the label because this is not confirmed as a bug in LuaRocks. Crashes during development as an extension can happen due to misuse of internal APIs (in that case it would be more similar to "user error" than a bug). To your other question about the procedure for testing these extensions, indeed we don't have a good user story for it — there hasn't been much interest in writing extensions historically, so support is pretty minimal at this time, unfortunately. These extensions points are provided in a best-effort basis, but sadly we don't have a documented stable programmatic API or testing tooling to point you to.
> 
> If you require `luarocks.*` modules from your extension and pass invalid arguments to functions, you will get crashes, and LuaRocks will say "please report a bug" along with the stacktrace because that's the default behavior of the custom error handler designed to catch LuaRocks internal bugs — I agree that's misleading, but there's not much we can do there.

I understand, thank you

Do you know whats going on then?

https://github.com/Frityet/Fuser/blob/011a2381ba93228c75d744a2e4d28547195d2592/src/fuser.lua
